File: example_scripts/rdf_parser.py
"""
RiverWare RDF file parser.

RDF format (text, one token per line):
  Package preamble  key:value ... END_PACKAGE_PREAMBLE
  Per run:
    Run preamble    key:value ... END_RUN_PREAMBLE
    Timestep list   YYYY-M-D 24:00  (time_steps lines)
    Per slot:
      Slot preamble key:value ... END_SLOT_PREAMBLE
      units line
      scale line
      values        (time_steps lines for series; 1 line for scalar)
      END_COLUMN
      END_SLOT
    END_RUN
"""

from __future__ import annotations

import sys
from datetime import datetime, timedelta
from pathlib import Path


def _normalize_ts(raw: str) -> str:
    """Convert 'YYYY-M-D 24:00' → ISO date 'YYYY-MM-DD'."""
    date_part = raw.split()[0]
    dt = datetime.strptime(date_part, "%Y-%m-%d")
    # RiverWare 24:00 = end of that calendar day — keep as-is
    return dt.strftime("%Y-%m-%d")


def _parse_preamble(lines: list[str], pos: int, end_marker: str) -> tuple[dict, int]:
    """
    Read key:value lines from pos until end_marker (exclusive).
    Returns (dict, new_pos).  end_marker line is consumed.
    If end_marker == 1 (int), reads exactly one line.
    """
    result: dict = {}
    while pos < len(lines):
        line = lines[pos]
        pos += 1
        if line == end_marker:
            break
        parts = line.split(":", 1)
        key = parts[0].strip()
        value = parts[1].strip() if len(parts) > 1 else None
        result[key] = value
        if end_marker == 1:  # sentinel: read exactly one line
            break
    return result, pos


def _parse_slot(lines: list[str], pos: int, nts: int) -> tuple[dict, int]:
    """Parse one slot block starting at pos (first line = object_type or similar)."""
    slot, pos = _parse_preamble(lines, pos, "END_SLOT_PREAMBLE")

    # units and scale are bare single-value lines (no key prefix)
    units_line = lines[pos]; pos += 1
    scale_line = lines[pos]; pos += 1

    # Parse bare "key: value" or just "value"
    def _bare_value(line: str) -> str:
        return line.split(":", 1)[-1].strip() if ":" in line else line.strip()

    slot["units"] = _bare_value(units_line)
    slot["scale"] = _bare_value(scale_line)

    # Find next END_COLUMN to determine scalar vs series
    ec_idx = next(
        (i for i in range(pos, len(lines)) if lines[i] == "END_COLUMN"),
        None,
    )
    if ec_idx is None:
        raise ValueError(f"END_COLUMN not found after position {pos}")

    n_values = ec_idx - pos
    slot_type = (slot.get("slot_type") or "").strip().lower()
    is_series_slot = "series" in slot_type
    is_scalar_slot = "scalar" in slot_type

    if is_series_slot:
        slot["scalar"] = False
    elif is_scalar_slot:
        slot["scalar"] = True
    elif n_values == nts:
        slot["scalar"] = False
    elif n_values == 1:
        slot["scalar"] = True
    else:
        import warnings
        warnings.warn(
            f"Slot '{slot.get('object_name')}.{slot.get('slot_name')}': "
            f"expected {nts} or 1 values, found {n_values}. Skipping."
        )
        slot["values"] = []
        slot["scalar"] = None
        pos = ec_idx + 1  # skip to after END_COLUMN
        if pos < len(lines) and lines[pos] == "END_SLOT":
            pos += 1
        return slot, pos

    raw_values = lines[pos : pos + n_values]
    try:
        slot["values"] = [float(v) for v in raw_values]
    except ValueError:
        slot["values"] = raw_values  # keep as strings if not numeric

    pos = ec_idx + 1  # advance past END_COLUMN
    if pos < len(lines) and lines[pos] == "END_SLOT":
        pos += 1

    return slot, pos


def _parse_run(lines: list[str], pos: int) -> tuple[dict, int]:
    """Parse one run block.  pos points to first line after END_PACKAGE_PREAMBLE
    (or after previous END_RUN)."""
    run: dict = {}

    preamble, pos = _parse_preamble(lines, pos, "END_RUN_PREAMBLE")
    run["preamble"] = preamble

    nts = int(preamble.get("time_steps") or preamble.get("timesteps") or 0)

    # Read timestep list
    raw_times = lines[pos : pos + nts]
    run["times"] = [_normalize_ts(t) for t in raw_times]
    pos += nts

    # Read slots until END_RUN
    run["slots"] = {}
    while pos < len(lines):
        if lines[pos] == "END_RUN":
            pos += 1
            break

        slot, pos = _parse_slot(lines, pos, nts)
        key = f"{slot.get('object_name', '')}.{slot.get('slot_name', '')}"
        run["slots"][key] = slot

    return run, pos


def parse_rdf(path: str | Path) -> dict:
    """
    Parse a RiverWare RDF file.

    Returns:
        {
            "meta": {key: value, ...},
            "runs": [
                {
                    "preamble": {key: value, ...},
                    "times": ["YYYY-MM-DD", ...],
                    "slots": {
                        "ObjectName.SlotName": {
                            "object_type": ..., "object_name": ...,
                            "slot_type": ..., "slot_name": ...,
                            "units": ..., "scale": ...,
                            "scalar": bool,
                            "values": [float, ...]
                        },
                        ...
                    }
                },
                ...
            ]
        }
    """
    path = Path(path)
    if path.suffix.lower() != ".rdf":
        raise ValueError(f"{path} does not appear to be an rdf file.")
    if not path.exists():
        raise FileNotFoundError(f"{path} does not exist.")

    lines = path.read_text(encoding="utf-8", errors="replace").splitlines()
    # Strip carriage returns from Windows line endings
    lines = [ln.rstrip("\r") for ln in lines]

    pos = 0
    meta, pos = _parse_preamble(lines, pos, "END_PACKAGE_PREAMBLE")

    expected_runs = int(meta.get("number_of_runs") or 0)
    runs: list[dict] = []

    while len(runs) < expected_runs and pos < len(lines):
        run, pos = _parse_run(lines, pos)
        runs.append(run)

    return {"meta": meta, "runs": runs}


def list_slots(rdf: dict) -> list[dict]:
    """Return slot metadata list from first run (all runs share the same slots)."""
    if not rdf["runs"]:
        return []
    return [
        {
            "key": key,
            "object_name": s.get("object_name", ""),
            "slot_name": s.get("slot_name", ""),
            "object_type": s.get("object_type", ""),
            "slot_type": s.get("slot_type", ""),
            "units": s.get("units", ""),
            "scale": s.get("scale", ""),
            "scalar": s.get("scalar"),
        }
        for key, s in rdf["runs"][0]["slots"].items()
    ]

File: src/rdf_parser.cpp
#include "rdf2parquet/rdf_parser.hpp"

#include <algorithm>
#include <cctype>
#include <cmath>
#include <optional>
#include <sstream>

namespace rdf2parquet {

namespace {

// --- small string helpers ---------------------------------------------------

std::string_view trim(std::string_view s) {
  const auto isSpace = [](unsigned char c) { return std::isspace(c) != 0; };
  while (!s.empty() && isSpace(static_cast<unsigned char>(s.front()))) s.remove_prefix(1);
  while (!s.empty() && isSpace(static_cast<unsigned char>(s.back()))) s.remove_suffix(1);
  return s;
}

std::string toLower(std::string_view s) {
  std::string out(s);
  std::transform(out.begin(), out.end(), out.begin(),
                  [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
  return out;
}

std::vector<std::string> splitLines(std::string_view text) {
  std::vector<std::string> lines;
  std::size_t start = 0;
  while (start <= text.size()) {
    std::size_t nl = text.find('\n', start);
    std::string_view line =
        (nl == std::string_view::npos) ? text.substr(start) : text.substr(start, nl - start);
    if (!line.empty() && line.back() == '\r') line.remove_suffix(1);
    lines.emplace_back(line);
    if (nl == std::string_view::npos) break;
    start = nl + 1;
  }
  return lines;
}

// Bare "key: value" or plain "value" line -> the value portion, trimmed.
std::string bareValue(const std::string& line) {
  auto idx = line.find(':');
  if (idx == std::string::npos) return std::string(trim(line));
  return std::string(trim(std::string_view(line).substr(idx + 1)));
}

// Coerce a token to a finite double, or nullopt if not finite/not numeric
// (REQ-001: per-value coercion).
std::optional<double> safeFloat(const std::string& token) {
  std::string_view t = trim(token);
  if (t.empty()) return std::nullopt;
  try {
    std::size_t idx = 0;
    double v = std::stod(std::string(t), &idx);
    if (idx != t.size() || !std::isfinite(v)) return std::nullopt;
    return v;
  } catch (...) {
    return std::nullopt;
  }
}

// Parse a fully-numeric integer token, or nullopt (used for `trace` and
// count preamble fields; a partially-numeric token like "12a" is rejected
// rather than truncated).
std::optional<long> tryParseInt(const std::string& token) {
  std::string_view t = trim(token);
  if (t.empty()) return std::nullopt;
  try {
    std::size_t idx = 0;
    long v = std::stol(std::string(t), &idx);
    if (idx != t.size()) return std::nullopt;
    return v;
  } catch (...) {
    return std::nullopt;
  }
}

// Convert 'YYYY-M-D 24:00' -> ISO date 'YYYY-MM-DD'. RiverWare's 24:00 means
// end-of-day; the calendar date is kept as-is (matches the Python reference
// _normalize_ts). Single-digit month/day are zero-padded.
std::string normalizeTimestamp(const std::string& raw, std::size_t lineNo) {
  std::string_view trimmed = trim(raw);
  std::size_t spaceIdx = trimmed.find_first_of(" \t");
  std::string_view datePart =
      (spaceIdx == std::string_view::npos) ? trimmed : trimmed.substr(0, spaceIdx);

  // Expect \d{4}-\d{1,2}-\d{1,2}
  auto isDigit = [](char c) { return c >= '0' && c <= '9'; };
  std::size_t i = 0;
  auto readDigits = [&](std::size_t maxLen) -> std::string {
    std::size_t j = i;
    while (j < datePart.size() && j - i < maxLen && isDigit(datePart[j])) ++j;
    std::string out(datePart.substr(i, j - i));
    i = j;
    return out;
  };

  std::string year = readDigits(4);
  bool ok = year.size() == 4 && i < datePart.size() && datePart[i] == '-';
  std::string month, day;
  if (ok) {
    ++i;
    month = readDigits(2);
    ok = !month.empty() && month.size() <= 2 && i < datePart.size() && datePart[i] == '-';
  }
  if (ok) {
    ++i;
    day = readDigits(2);
    ok = !day.empty() && day.size() <= 2 && i == datePart.size();
  }
  if (!ok) {
    throw RdfParseError("Malformed RDF timestamp: \"" + raw + "\"", lineNo);
  }
  if (month.size() == 1) month = "0" + month;
  if (day.size() == 1) day = "0" + day;
  return year + "-" + month + "-" + day;
}

// Read "key:value" lines from pos until endMarker (exclusive); endMarker
// line is consumed. Throws if EOF is reached without finding endMarker
// (REQ-006: missing END markers are a structural, exit-2 error).
std::map<std::string, std::string> parsePreamble(const std::vector<std::string>& lines,
                                                   std::size_t& pos, const std::string& endMarker) {
  std::map<std::string, std::string> result;
  while (pos < lines.size()) {
    const std::string& line = lines[pos];
    ++pos;
    if (line == endMarker) return result;
    auto idx = line.find(':');
    if (idx == std::string::npos) {
      result[std::string(trim(line))] = "";
    } else {
      std::string key(trim(std::string_view(line).substr(0, idx)));
      std::string value(trim(std::string_view(line).substr(idx + 1)));
      result[key] = value;
    }
  }
  throw RdfParseError("Unexpected end of file: missing " + endMarker, pos);
}

struct SlotResult {
  RdfSlot slot;
  std::size_t pos;
};

SlotResult parseSlot(const std::vector<std::string>& lines, std::size_t pos, long nts,
                      std::vector<std::string>& warnings) {
  auto preamble = parsePreamble(lines, pos, "END_SLOT_PREAMBLE");

  if (pos + 1 >= lines.size()) {
    throw RdfParseError("Unexpected end of file: missing units/scale lines for a slot.", pos + 1);
  }
  const std::string& unitsLine = lines[pos]; ++pos;
  const std::string& scaleLine = lines[pos]; ++pos;

  RdfSlot slot;
  slot.object_type = preamble.count("object_type") ? preamble["object_type"] : "";
  slot.object_name = preamble.count("object_name") ? preamble["object_name"] : "";
  slot.slot_type = preamble.count("slot_type") ? preamble["slot_type"] : "";
  slot.slot_name = preamble.count("slot_name") ? preamble["slot_name"] : "";
  slot.units = bareValue(unitsLine);
  slot.scale = bareValue(scaleLine);

  std::size_t ecIdx = pos;
  while (ecIdx < lines.size() && lines[ecIdx] != "END_COLUMN") ++ecIdx;
  if (ecIdx >= lines.size()) {
    throw RdfParseError("END_COLUMN not found after position " + std::to_string(pos + 1),
                         pos + 1);
  }

  long nValues = static_cast<long>(ecIdx - pos);
  std::string slotType = toLower(slot.slot_type);
  bool isSeriesSlot = slotType.find("series") != std::string::npos;
  bool isScalarSlot = slotType.find("scalar") != std::string::npos;

  bool matched = true;
  if (isSeriesSlot) {
    slot.is_scalar = false;
  } else if (isScalarSlot) {
    slot.is_scalar = true;
  } else if (nValues == nts) {
    slot.is_scalar = false;
  } else if (nValues == 1) {
    slot.is_scalar = true;
  } else {
    matched = false;
  }

  if (!matched) {
    std::ostringstream msg;
    msg << "Slot '" << slot.object_name << "." << slot.slot_name << "': expected " << nts
        << " or 1 values, found " << nValues << ". Skipping.";
    warnings.push_back(msg.str());
    slot.is_scalar = std::nullopt;
  } else {
    for (std::size_t i = pos; i < pos + static_cast<std::size_t>(nValues); ++i) {
      slot.values.push_back(safeFloat(lines[i]));
    }
  }

  pos = ecIdx + 1;
  // REQ-006 deviation from the Python/JS references: the line after
  // END_COLUMN MUST be END_SLOT. A multi-column slot (more than one value
  // column) puts something else here; the references silently mis-parse
  // that as the start of the next slot's preamble, so we reject it instead.
  if (pos >= lines.size() || lines[pos] != "END_SLOT") {
    throw RdfParseError(
        "Expected END_SLOT after END_COLUMN for slot '" + slot.object_name + "." +
            slot.slot_name + "' (multi-column slots are not supported).",
        pos + 1);
  }
  ++pos;

  return SlotResult{std::move(slot), pos};
}

struct RunResult {
  RdfRun run;
  std::size_t pos;
};

RunResult parseRun(const std::vector<std::string>& lines, std::size_t pos, int ordinal,
                    std::vector<std::string>& warnings) {
  RdfRun run;
  run.preamble = parsePreamble(lines, pos, "END_RUN_PREAMBLE");

  auto timeStepsIt = run.preamble.find("time_steps");
  if (timeStepsIt == run.preamble.end()) timeStepsIt = run.preamble.find("timesteps");
  long nts = 0;
  if (timeStepsIt != run.preamble.end()) {
    auto parsed = tryParseInt(timeStepsIt->second);
    if (!parsed || *parsed < 0) {
      throw RdfParseError("Invalid time_steps value: \"" + timeStepsIt->second + "\"", pos);
    }
    nts = *parsed;
  }
  // SEC-001: reject counts that would over-allocate before touching them.
  if (static_cast<std::size_t>(nts) > lines.size() - pos) {
    throw RdfParseError(
        "time_steps (" + std::to_string(nts) + ") exceeds remaining lines in file.", pos);
  }

  auto traceIt = run.preamble.find("trace");
  if (traceIt != run.preamble.end()) {
    auto parsed = tryParseInt(traceIt->second);
    run.trace_id = parsed ? static_cast<int32_t>(*parsed) : ordinal;
  } else {
    run.trace_id = ordinal;
  }

  run.timesteps.reserve(static_cast<std::size_t>(nts));
  for (long i = 0; i < nts; ++i) {
    run.timesteps.push_back(normalizeTimestamp(lines[pos + static_cast<std::size_t>(i)], pos + 1));
  }
  pos += static_cast<std::size_t>(nts);

  bool sawEndRun = false;
  while (pos < lines.size()) {
    if (lines[pos] == "END_RUN") {
      ++pos;
      sawEndRun = true;
      break;
    }
    auto slotResult = parseSlot(lines, pos, nts, warnings);
    pos = slotResult.pos;
    run.slots.push_back(std::move(slotResult.slot));
  }
  if (!sawEndRun) {
    throw RdfParseError("Unexpected end of file: missing END_RUN.", pos);
  }

  return RunResult{std::move(run), pos};
}

}  // namespace

RdfParseError::RdfParseError(const std::string& message, std::size_t line)
    : std::runtime_error(line == 0 ? message : message + " (line " + std::to_string(line) + ")"),
      line_(line) {}

RdfFile parseRdf(std::string_view text) {
  std::vector<std::string> lines = splitLines(text);

  RdfFile file;
  std::size_t pos = 0;
  file.package_preamble = parsePreamble(lines, pos, "END_PACKAGE_PREAMBLE");

  auto runsIt = file.package_preamble.find("number_of_runs");
  if (runsIt == file.package_preamble.end()) {
    throw RdfParseError("Not a valid RDF file: missing number_of_runs.", pos);
  }
  auto expectedRuns = tryParseInt(runsIt->second);
  if (!expectedRuns || *expectedRuns < 0) {
    throw RdfParseError("Invalid number_of_runs value: \"" + runsIt->second + "\"", pos);
  }
  // SEC-001: a run needs at least an END_RUN_PREAMBLE and END_RUN line, so
  // this is a safe lower-bound sanity cap against a corrupt/huge count.
  if (static_cast<std::size_t>(*expectedRuns) > lines.size() - pos) {
    throw RdfParseError(
        "number_of_runs (" + std::to_string(*expectedRuns) + ") exceeds remaining lines in file.",
        pos);
  }

  while (file.runs.size() < static_cast<std::size_t>(*expectedRuns) && pos < lines.size()) {
    auto result = parseRun(lines, pos, static_cast<int>(file.runs.size()) + 1, file.warnings);
    pos = result.pos;
    file.runs.push_back(std::move(result.run));
  }

  if (file.runs.empty()) {
    throw RdfParseError("RDF file contains no runs.", pos);
  }

  return file;
}

}  // namespace rdf2parquet
